li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -0,0 +1,274 @@
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This is a producer implementation that works with a federated Kafka cluster, which consists of one or more physical
+ * Kafka clusters.
+ */
+public class LiKafkaFederatedProducerImpl<K, V> implements LiKafkaProducer<K, V> {
+  private static final Logger LOG = LoggerFactory.getLogger(LiKafkaFederatedProducerImpl.class);
+
+  // Per-cluster producer context
+  private class PerClusterProducerContext<K, V> {
+    private Producer<byte[], byte[]> _producer;
+    /*package private for testing*/ Auditor<K, V> _auditor;
+
+    // A counter of the threads in the middle of sending messages. This is needed to ensure when we close the producer
+    // everything is audited.
+    private final AtomicInteger _numThreadsInSend = new AtomicInteger();
+    private volatile boolean _closed;
+
+    // This is null if the underlying producer does not have an implementation for time-bounded flush
+    private final Method _boundedFlushMethod;
+    private final AtomicInteger _boundFlushThreadCount = new AtomicInteger();
+
+    public PerClusterProducerContext(LiKafkaProducerConfig configs) {
+      _producer = new KafkaProducer<>(configs.originals(), new ByteArraySerializer(), new ByteArraySerializer());
+
+      // TODO: This hack remains until bounded flush is added to apache/kafka
+      Method producerSupportsBoundedFlush;
+      try {
+        producerSupportsBoundedFlush = _producer.getClass().getMethod("flush", long.class, TimeUnit.class);
+      } catch (NoSuchMethodException e) {
+        LOG.warn("Wrapped KafkaProducer does not support time-bounded flush.", e);
+        producerSupportsBoundedFlush = null;
+      }
+      _boundedFlushMethod = producerSupportsBoundedFlush;
+
+      // TODO: add an option to pass custom auditor
+      _auditor = configs.getConfiguredInstance(LiKafkaProducerConfig.AUDITOR_CLASS_CONFIG, Auditor.class, _producer);
+      _auditor.start();
+
+      _closed = false;
+    }
+
+    public Producer<byte[], byte[]> producer() {
+      return _producer;
+    }
+
+    public Auditor<K, V> auditor() {
+      return _auditor;
+    }
+
+    public AtomicInteger numThreadsInSend() {
+      return _numThreadsInSend;
+    }
+
+    public boolean closed() {
+      return _closed;
+    }
+  }
+
+  // The cluster group this client is talking to
+  private ClusterGroupDescriptor _clusterGroup;
+
+  // The client for the metadata service which serves cluster and topic metadata
+  private MetadataServiceClient _mdsClient;
+
+  // The id of this client assigned by the metadata service
+  private UUID _clientId;
+
+  // Per cluster producers, which sends raw bytes
+  private Map<String, PerClusterProducerContext<K, V>> _producers;
```

Comment:
just like we have a ClusterGroupDescriptor, do we want to have a ClusterDescriptor? could have a name, bootstrap url, maybe zk url etc ?

Reply:
I actually have that in my next PR... moved it here.